### PR TITLE
planner: fix wrong TableDual plans caused by comparing Binary and Bytes incorrectly (#23860)

### LINKED ISSUE #23846: The execution result of using TableDual plan is not correct
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
create table t(a varbinary(10),UNIQUE KEY(a));
insert into t values(0x00A4EEF4FA55D6706ED5);
select count(*) from t where a=0x00A4EEF4FA55D6706ED5; -- 1
desc select * from t where a=0x00A4EEF4FA55D6706ED5; -- TableDual_2 plan
select * from t where a=0x00A4EEF4FA55D6706ED5; -- Empty result
```
### 2. What did you expect to see? (Required)
0x00A4EEF4FA55D6706ED5
### 3. What did you see instead (Required)
Empty result
### 4. What is your TiDB version? (Required)

Release Version: v4.0.0-beta.2-2561-g9c75cfa4e
Edition: Community
Git Commit Hash: 9c75cfa4e2bbb854ab90562ecec0409f84a989b0
Git Branch: master
UTC Build Time: 2021-04-05 07:39:46
GoVersion: go1.15.1
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #23860 to release-4.0
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/23917
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/23917:release-4.0-af0982805940
```

---

### What problem does this PR solve?

Issue Number: close #23846 

Problem Summary: planner: fix wrong TableDual plans caused by comparing Binary and Bytes incorrectly

### What is changed and how it works?

planner: fix wrong TableDual plans caused by comparing Binary and Bytes incorrectly

### Check List 

Tests 

- Unit test

### Release note 

- planner: fix wrong TableDual plans caused by comparing Binary and Bytes incorrectly

### CODE REVIEW

File: planner/core/integration_test.go [hunk 1]
```diff
@@ -1934,3 +1934,76 @@ func (s *testIntegrationSuite) TestGetVarExprWithBitLiteral(c *C) {
 	tk.MustExec("set @a = 0b11000100110101;")
 	tk.MustQuery("execute stmt using @a;").Check(testkit.Rows("1"))
 }
+<<<<<<< HEAD
+=======
```
Comment (eurekaka) [inline planner/core/integration_test.go:1938]: @qw4990 please resolve conflicts.

**Comment by qw4990** [comment]
Wait for https://github.com/pingcap/tidb/pull/23655

**Review by rebelice — APPROVED** [review]
/lgtm

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**